### PR TITLE
[core] memory monitor to use the event loop from node manager

### PR DESCRIPTION
**Pull request by clarng**
## Why are these changes needed?

The objects accessed by the memory monitor callback are not thread-safe. The node manager runs on a specific event loop and this PR changes the memory monitor to share that same event loop similar to the rest of the components executed under the node manager.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/raylet/node_manager.cc [hunk 1]
```diff
@@ -2969,10 +2969,14 @@ MemoryUsageRefreshCallback NodeManager::CreateMemoryUsageRefreshCallback() {
           /// TODO: (clarng) right now destroy is called after the messages are created
           /// since we print the process memory in the message. Destroy should be called
           /// as soon as possible to free up memory.
-          DestroyWorker(high_memory_eviction_target_,
-                        rpc::WorkerExitType::USER_ERROR,
-                        worker_exit_message,
-                        true /* force */);
+          this->io_service_.post(
```
Comment (rkooo567) [inline src/ray/raylet/node_manager.cc:2972]: Why don't we put the whole code into this service? There are still potential races. E.g., accessing `        auto workers = this->WorkersWithLatestSubmittedTasks();
`

Comment (clarng) [reply src/ray/raylet/node_manager.cc]: agree - looks like the subcomponents of the node manager are sharing the event loop, seems more consistent to have the memory monitor be the same

**Comment by clarng** [comment]
Will share the release test link once we have a wheel

**Comment by clarng** [comment]
moved to https://github.com/ray-project/ray/pull/28271

**Pull request closed without merging**